staging/vhost-device-can/src/vhu_can.rs

```diff
@@ -0,0 +1,1224 @@
+                    trace!("VIRTIO_CAN_SET_CTRL_MODE_START");
+                    Ok(())
+                }
+                VIRTIO_CAN_SET_CTRL_MODE_STOP => {
+                    //TODO: vcan->busoff = false;
+                    trace!("VIRTIO_CAN_SET_CTRL_MODE_STOP");
+                    Ok(())
+                }
+                _ => {
+                    trace!("Ctrl queue: msg type 0x{:?} unknown", request.msg_type);
+                    return Err(Error::HandleEventUnknown);
+                }
+            }?;
+
+            let desc_response = descriptors[1];
+            if !desc_response.is_write_only() {
+                return Err(Error::UnexpectedReadableDescriptor(1));
+            }
+
+            let response = VIRTIO_CAN_RESULT_OK;
+
+            desc_chain
+                .memory()
+                .write_slice(response.as_slice(), desc_response.addr())
+                .map_err(|_| Error::DescriptorWriteFailed)?;
+
+            if vring
+                .add_used(desc_chain.head_index(), desc_response.len())
+                .is_err()
+            {
+                warn!("Couldn't return used descriptors to the ring");
+            }
+        }
+
+        Ok(true)
+    }
+
+    fn process_tx_requests(
+        &self,
+        requests: Vec<CanDescriptorChain>,
+        vring: &VringRwLock,
+    ) -> Result<bool> {
+        trace!("process_tx_requests");
+
+        if requests.is_empty() {
+            trace!("requests.is_empty()");
+            return Ok(true);
+        }
+
+        for desc_chain in requests {
+            let descriptors: Vec<_> = desc_chain.clone().collect();
+
+            if descriptors.len() != 2 {
+                return Err(Error::UnexpectedDescriptorCount(descriptors.len()));
+            }
+
+            let desc_request = descriptors[0];
+            if desc_request.is_write_only() {
+                return Err(Error::UnexpectedWriteOnlyDescriptor(0));
+            }
+
+            if (desc_request.len() as usize) > size_of::<VirtioCanFrame>()
+                || (desc_request.len() as usize) < (size_of::<VirtioCanFrame>() - 64)
+            {
+                trace!(
+                    "desc_len: {}, virtio_can_len: {:?}\n",
+                    desc_request.len() as usize,
+                    size_of::<VirtioCanFrame>()
+                );
+                return Err(Error::UnexpectedTxDescriptorSize(
+                    size_of::<VirtioCanFrame>() - 64,
+                    size_of::<VirtioCanFrame>(),
+                    desc_request.len(),
+                ));
+            }
+
+            let request = desc_chain
+                .memory()
+                .read_obj::<VirtioCanFrame>(desc_request.addr())
+                .map_err(|_| Error::DescriptorReadFailed)?;
+
+            CanController::print_can_frame(request);
+
+            let msg_type = request.msg_type.to_native();
+            let mut can_id = request.can_id.to_native();
```

Comment:
Based on the spec, the device shall respond with NOT_OK if the can_id is out-of-range. I think that check is missed here.

Reply:
Yes, but I am not 100% sure what does this mean. My best interpretation would be that, if the received frame does not have `CAN_EFF_FLAG` under `can_id` and the `can_id & CAN_EFF_MASK > CAN_SFF_MASK` then return NOT_OK.

As far as I see in Opensynergy QEMU virtio-can device implementation ([tranmission](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L153), [receiption](https://github.com/OpenSynergy/qemu/blob/115540168f92ba5351a20b9c62552782ea1e3e04/hw/net/can/can_virtio.c#L280)):
```
    // Transmission
    if (frame->can_id & QEMU_CAN_EFF_FLAG) {
        flags = VIRTIO_CAN_FLAGS_EXTENDED;
        can_rx.can_id = cpu_to_le32(frame->can_id & QEMU_CAN_EFF_MASK);
    } else {
        flags = 0;
        can_rx.can_id = cpu_to_le32(frame->can_id & QEMU_CAN_SFF_MASK);
    }
```
The code does not return any error, instead apply `QEMU_CAN_SFF_MASK` if `QEMU_CAN_EFF_FLAG` is in place (this is how that case is implemented in vhost-device-can).
So please let me know if my interpetation of the specs is correct, and if you like to change the behavior according to it.

Reply:
I just submitted this patch to explain that in the spec: https://lists.oasis-open.org/archives/virtio-comment/202402/msg00102.html

Reply:
Thank's Matias, that's been addressed into the code.